### PR TITLE
Implemented SchemaTransforms for SingleStoreIO

### PR DESCRIPTION
**Pull request by AdalbertMemSQL**
Added default RowMapper and UserDataMapper.
Implemented SchemaTransform for Read, ReadWithPartitions, and Write PTransforms.
These changes will allow us to configure SingleStoreIO easier and to use it with other languages

**Please** add a meaningful description for your change here

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by AdalbertMemSQL** [comment]
R: @Abacn

**Comment by johnjcasey** [comment]
r: @ahmedabu98

**Comment by johnjcasey** [comment]
@AdalbertMemSQL it is probably worth putting some of these classes into a subdirectory (singlestore/schematransform) to help with organization

File: sdks/java/io/singlestore/src/main/java/org/apache/beam/sdk/io/singlestore/schematransform/SingleStoreSchemaTransformReadWithPartitionsProvider.java
Comment (ahmedabu98) [reply sdks/java/io/singlestore/src/main/java/org/apache/beam/sdk/io/singlestore/schematransform/SingleStoreSchemaTransformReadWithPartitionsProvider.java:35]: Some thought should be put into this decision. Merging the two read modes would make sense as it is now, but if it's likely that these two modes will develop down the line to have many more differences then keeping them separate makes more sense.

File: sdks/java/io/singlestore/src/main/java/org/apache/beam/sdk/io/singlestore/schematransform/SingleStoreSchemaTransformWriteProvider.java
Comment (ahmedabu98) [inline sdks/java/io/singlestore/src/main/java/org/apache/beam/sdk/io/singlestore/schematransform/SingleStoreSchemaTransformWriteProvider.java:60]: See previous comment for URN conventions

**Review by ahmedabu98 — CHANGES_REQUESTED** [review]
Thank you, I left a few comments. This IO looks great!

File: sdks/java/io/singlestore/src/main/java/org/apache/beam/sdk/io/singlestore/schematransform/SingleStoreSchemaTransformReadWithPartitionsProvider.java
Comment (AdalbertMemSQL) [reply sdks/java/io/singlestore/src/main/java/org/apache/beam/sdk/io/singlestore/schematransform/SingleStoreSchemaTransformReadWithPartitionsProvider.java:35]: initialNumReaders parameter is already deleted. So now the only difference is outputParallelization parameter that has sense only for the sequential reading. I don't think that these read modes will evolve a lot. Will try to merge their SchemaTransforms.

File: sdks/java/io/singlestore/src/main/java/org/apache/beam/sdk/io/singlestore/schematransform/SingleStoreSchemaTransformReadConfiguration.java
Comment (AdalbertMemSQL) [reply sdks/java/io/singlestore/src/main/java/org/apache/beam/sdk/io/singlestore/schematransform/SingleStoreSchemaTransformReadConfiguration.java:69]: Hmm...
That's a good question.
Is it correct that if I will `@DefaultSchema(AutoValueSchema.class)` before the `DataSourceConfiguration` class then Beam will infer the schema for it and an object with the same schema can be somehow created in other SDKs?

File: sdks/java/io/singlestore/src/main/java/org/apache/beam/sdk/io/singlestore/schematransform/SingleStoreSchemaTransformReadProvider.java
Comment (adalbert44) [reply sdks/java/io/singlestore/src/main/java/org/apache/beam/sdk/io/singlestore/schematransform/SingleStoreSchemaTransformReadProvider.java:128]: spootbug doesn't recognize this check 
```
/home/amakarovych-ua/projects/beam/sdks/java/io/singlestore/src/main/java/org/apache/beam/sdk/io/singlestore/SingleStoreIO.java:343: error: [argument] incompatible argument for parameter password of setPassword.
        basicDataSource.setPassword(password);
                                    ^
  found   : @Initialized @Nullable String
  required: @Initialized @NonNull String
```

I will use `table != null && !table.isEmpty()`

**Comment by AdalbertMemSQL** [comment]
@ahmedabu98 Can you please do one more review of this PR?

**Comment by ahmedabu98** [comment]
Sorry for the holdup, taking a look now

File: .test-infra/jenkins/job_PostCommit_Java_SingleStoreIO_IT.groovy
Comment (ahmedabu98) [inline .test-infra/jenkins/job_PostCommit_Java_SingleStoreIO_IT.groovy:44]: Can you set a timeout here? This is not to set a strict time limit for the job, but more to catch runaway jobs, see example [here](https://github.com/apache/beam/blob/ef5351ad50a817498ea9e34a7c514dd9d60fb143/.test-infra/jenkins/job_PostCommit_Java.groovy#L32).

File: sdks/java/io/singlestore/src/main/java/org/apache/beam/sdk/io/singlestore/SingleStoreDefaultUserDataMapper.java
Comment (ahmedabu98) [inline sdks/java/io/singlestore/src/main/java/org/apache/beam/sdk/io/singlestore/SingleStoreDefaultUserDataMapper.java:40]: You can make use of [`FieldType::isLogicalType`](https://github.com/apache/beam/blob/3ffaa49c3217386291fa163d0a868a686b329ac6/sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/Schema.java#L643) as a check here.

File: sdks/java/io/singlestore/src/main/java/org/apache/beam/sdk/io/singlestore/schematransform/SingleStoreSchemaTransformReadConfiguration.java
Comment (ahmedabu98) [reply sdks/java/io/singlestore/src/main/java/org/apache/beam/sdk/io/singlestore/schematransform/SingleStoreSchemaTransformReadConfiguration.java:72]: Just tested this out locally with a simple configuration that had a POJO field and generating a schema worked fine with just the `@AutoValue` decoration: 
```
Field{name=field1, description=, type=STRING NOT NULL, options={{}}}
Field{name=field2, description=, type=INT32 NOT NULL, options={{}}}
Field{name=pojoField, description=, type=ROW<
    pojoField1 STRING NOT NULL, 
    pojoField2 INT32 NOT NULL
  > NOT NULL, options={{}}}
```

Comment (ahmedabu98) [reply sdks/java/io/singlestore/src/main/java/org/apache/beam/sdk/io/singlestore/schematransform/SingleStoreSchemaTransformReadConfiguration.java:72]: It might still be better to keep `@DefaultSchema(AutoValueSchema.class)` though, according to the programming guide: https://beam.apache.org/documentation/programming-guide/#inferring-schemas

File: .test-infra/jenkins/job_PostCommit_Java_SingleStoreIO_IT.groovy
Comment (AdalbertMemSQL) [reply .test-infra/jenkins/job_PostCommit_Java_SingleStoreIO_IT.groovy:44]: Looks like the default timeout is already set to 100
```
  // Sets common top-level job properties for main repository jobs.
  static void setTopLevelMainJobProperties(def context,
      String defaultBranch = 'master',
      int defaultTimeout = 100,
      boolean allowRemotePoll = true,
      String jenkinsExecutorLabel = 'beam',
      boolean cleanWorkspace = true) {

```

**Comment by ahmedabu98** [comment]
Thanks @AdalbertMemSQL, LGTM so far.

I had another design question, is it necessary to create two new read nested classes (`ReadRows` and `ReadWithPartitionsRows`) that cater to Row objects? Much of the code seems to be boilerplate for `Read` and `ReadWithPartitions`, except only to add some parameters to specify Row object output. It works as it is now, but it's better to be concise and not duplicate code.

It may help to reduce this by creating a new readRows() function that calls on SingleStoreIO.read() and adding the specifications needed to output Rows. Same with partitions, a new readWithPartitionsRows() function. See this [example](https://github.com/apache/beam/blob/ef5351ad50a817498ea9e34a7c514dd9d60fb143/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java#L575) in BigQueryIO. The same could be applied here, passing in the relevant rowMapper and coder. Although this will probably need a check  at the end of Read/ReadWithPartitions expand() to see if we are reading Rows so that it can set row schema on the output PCollection.

WDYT?

**Comment by ahmedabu98** [comment]
FYI these suggestions ^ (if you decide to implement them) can also be done in a later PR. I don't see any blockers for merging this one, let me know what you decide

**Comment by AdalbertMemSQL** [comment]
I like this idea.
Will try to implement it.

**Comment by AdalbertMemSQL** [comment]
@ahmedabu98 Can you please trigger execution of the job_PostCommit_Java_SingleStoreIO_IT.groovy?

**Comment by ahmedabu98** [comment]
Run Java SingleStoreIO_IT

**Comment by ahmedabu98** [comment]
Seed job failed here: https://ci-beam.apache.org/job/beam_SeedJob/10792/console

Seeing `ERROR: (job_PostCommit_Java_SingleStoreIO_IT.groovy, line 44) No such property: commonJobProperties for class: javaposse.jobdsl.dsl.jobs.FreeStyleJob`

**Review by ahmedabu98 — CHANGES_REQUESTED** [review]
Looks really good! left one comment and a typo fix. The typo is preventing us from running a seed job on this PR, which would allow us to run SingleStoreIO_IT.

File: sdks/java/io/singlestore/src/main/java/org/apache/beam/sdk/io/singlestore/SingleStoreIO.java
Comment (AdalbertMemSQL) [reply sdks/java/io/singlestore/src/main/java/org/apache/beam/sdk/io/singlestore/SingleStoreIO.java]: Now they can do it using RowMapperWithCodder
https://github.com/apache/beam/pull/24290/files#diff-1ae81cb3e2f6d00213f38c6ebcee815cea50c993e9a4a0514b0d93f7837af0bcR237-R240

File: sdks/java/io/singlestore/src/main/java/org/apache/beam/sdk/io/singlestore/SingleStoreDefaultUserDataMapper.java
Comment (ahmedabu98) [inline sdks/java/io/singlestore/src/main/java/org/apache/beam/sdk/io/singlestore/SingleStoreDefaultUserDataMapper.java:40]:
```suggestion
    checkArgument(
            type.getTypeName().isLogicalType(),
            "<appropriate error message>");;
```

make sure you're importing 
`import static org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.Preconditions.checkArgument;`

**Comment by ahmedabu98** [comment]
Run Java SingleStoreIO_IT

**Comment by ahmedabu98** [comment]
Integration test failed due to problems connecting to jdbc: https://ci-beam.apache.org/job/beam_PostCommit_Java_SingleStoreIO_IT_PR/1/console

**Comment by lukecwik** [comment]
Run Seed Job

**Comment by ahmedabu98** [comment]
Run Java SingleStoreIO_IT

File: sdks/java/io/singlestore/src/main/java/org/apache/beam/sdk/io/singlestore/SingleStoreDefaultUserDataMapper.java
Comment (ahmedabu98) [inline sdks/java/io/singlestore/src/main/java/org/apache/beam/sdk/io/singlestore/SingleStoreDefaultUserDataMapper.java:42]: Could you replace it with an appropriate error message?

**Comment by ahmedabu98** [comment]
Run Java SingleStoreIO_IT

**Comment by ahmedabu98** [comment]
@AdalbertMemSQL test is running now

**Review by ahmedabu98 — APPROVED** [review]
LGTM and tests are passing :) when you're ready I can look for a committer to merge this

**Comment by AdalbertMemSQL** [comment]
I'm ready
I will be grateful if you could find a committer :)

**Comment by pabloem** [comment]
I'll be happy to merge once tests are green again

**Comment by pabloem** [comment]
alright merging! Thanks everyone. Very happy to get this in!

**Comment by Abacn** [comment]
Looks like this PR breaks SingleStoreIO performance test: https://ci-beam.apache.org/view/PerformanceTests/job/beam_PerformanceTests_SingleStoreIO/79/

because the test source file has renamed to SingleStoreIOPerformanceIT.java instead of SingleStoreIOITPerformance.java

**Pull request merged**
